keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,401 @@
+)
+
+type ExecutionPolicy struct {
+	// Order in which Jobs will be created.
+	ExecutionPolicyOrder ExecutionPolicyOption `json:"executionPolicyOrder"`
+
+	// After all ReplicatedJobs reach this status, the JobSet will create the next ReplicatedJobs.
+	ExecutionPolicyRule []ExecutionPolicyRule `json:"rules"`
+}
+
+// ExecutionPolicyRule represents the execution policy rule for Job sequence.
+type ExecutionPolicyRule struct {
+
+	// Names of the replicated Jobs that applied the status.
+	TargetReplicatedJobs []string `json:"targetReplicatedJobs"`
+
+	// Status the target ReplicatedJobs must reach before subsequent ReplicatedJobs begin executing.
+	ReplicatedJobsStatus ReplicatedJobsStatusOption `json:"replicatedJobsStatus"`
+}
+
+type ReplicatedJobsStatusOption string
+
+// TODO: What statuses do we want to support in the first version ?
+const (
+	ReadyStatus ReplicatedJobsStatusOption = "Ready"
+
+	SucceededStatus ReplicatedJobsStatusOption = "Succeeded"
+
+	FailedStatus ReplicatedJobsStatusOption = "Failed"
+)
+```
+
+### Implementation
+
+The JobSet operator will control the creation of ReplicatedJobs based on their status. When
+desired replicated Jobs reach the status defined in replicatedJobsStatus, the controller creates
+the next set of replicated Jobs.
+
+If JobSet is suspended the all ReplicatedJobs will be suspended and the Job sequence starts again.
+
+### Quota Management
+
+In the initial implementation of the ExecutionPolicy the resource quota will be calculated as
+sum of all ReplicatedJobs resources. Which means JobSet will be admitted by
+[Kueue](https://github.com/kubernetes-sigs/kueue) only when all resources are available for
+every ReplicatedJob within JobSet.
+
+That allows us to leverage the existing integration between Kueue and JobSet while using the
+ExecutionPolicy API.
+
+In the future versions we will discuss potential partial admission of JobSet by Kueue. For example,
+when compute resources are available for the first ReplicatedJob the JobSet can be dispatched by
+Kueue.
+
+### Defaulting/Validation
+
+- ExecutionPolicy is immutable.
+- ExecutionPolicyOrder of `AnyOrder` is default setting.
+- StartupPolicy should be equal to `AnyOrder` when ExecutionPolicy is used.
+  - The StartupPolicy API will be deprecated over the next few JobSet releases, since
+    ExecutionPolicy can be used with ready status of ReplicatedJobs.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Unit Tests
+
+- `controllers`: `01/19/2024` - `30.2%`
+
+We will create a new file called execution_policy for the functionality.
+
+#### Integration tests
+
+- Having 2 Jobs running in sequence with Succeeded condition.
+- Suspend JobSet with execution policy being set.
+- Restart the failed Job when execution policy is set in JobSet.
```

Comment:
What would be the intended behavior for this test case? Wouldn't this depend on the JobSet's failure policy?

Reply:
It will, but we need to make sure that if JobSet fails, we restart the every ReplicatedJobs in the job sequence.